### PR TITLE
Implement c10 ops needed for benchmark

### PR DESCRIPTION
**Pull request by smessmer**
Summary:
This implements a first set of c10 operators, namely the ones needed for the multithread predictor benchmark.

All implementations are CPU-only and experimental. They're not meant to be used in production.

They can be used, however, to test calling simple c10 MLPs from Caffe2 or PyTorch when working on these integration paths.

Differential Revision: D8811698

### CODE REVIEW

File: caffe2/core/dispatch/DispatchTable.h [hunk 1]
```diff
@@ -39,6 +39,7 @@ class ThreadsafeOperatorTable_ final {
       auto result = map_.emplace(std::forward<Key>(key), value);
       if (!result.second) {
         std::ostringstream msg;
+        using ::operator<<;
```
Comment (ezyang) [inline caffe2/core/dispatch/DispatchTable.h]: Wait, why?

File: caffe2/utils/SmallVector.h [hunk 2]
```diff
@@ -0,0 +1,968 @@
+//===- llvm/ADT/SmallVector.h - 'Normally small' vectors --------*- C++ -*-===//
```
Comment (ezyang) [inline caffe2/utils/SmallVector.h]: Now that we have a unified cmake, we shouldn't copy paste SmallVector again, right? There's already an implementation in ATen

File: caffe2/utils/ArrayRef.h [hunk 3]
```diff
@@ -0,0 +1,201 @@
+//===--- ArrayRef.h - Array Reference Wrapper -------------------*- C++ -*-===//
```
Comment (ezyang) [inline caffe2/utils/ArrayRef.h]: Ditto here

File: caffe2/utils/AlignOf.h [hunk 4]
```diff
@@ -0,0 +1,145 @@
+//===--- AlignOf.h - Portable calculation of type alignment -----*- C++ -*-===//
```
Comment (ezyang) [inline caffe2/utils/AlignOf.h]: Ditto here

File: caffe2/operators/experimental/c10/cpu/sigmoid_cross_entropy_with_logits_cpu.cc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+#include "caffe2/core/dispatch/KernelRegistration.h"
+#include "caffe2/utils/math.h"
+
+using caffe2::Tensor;
+using caffe2::CPUContext;
+using caffe2::TIndex;
+
+namespace caffe2 {
+namespace {
+inline float sigmoid_partition(float lgt) {
```
Comment (ezyang) [inline caffe2/operators/experimental/c10/cpu/sigmoid_cross_entropy_with_logits_cpu.cc]: Why did these definitions need to get copy pasted?

**Comment by smessmer** [comment]
I'll remove the copy pasta.
This is meant for merging. It is a good set of operators covering common use cases like inputs vs. arguments, having no state vs having state, having variable number of inputs, ...
All designs we want to experiment with in the future need to make sure they support these cases, so this is basically a test that they do. Also, this allows future integration work (say C10 into ATen) to be tested with actual operators to call. There will likely have to be many changes to C10OperatorWrapper when we change to boxed, but probably fewer to the kernels themselves.
These operators are btw not meant to be used in any production code, that's why they're in the experimental directory and have such weird operator names. There's for example only CPU kernels currently and even on CPU, they might not support all argument combinations that are supported by caffe2.

**Pull request closed without merging**